### PR TITLE
add section about implied bounds

### PR DESCRIPTION
**Pull request by lcnr**
cc https://github.com/rust-lang/rust/pull/100747/files#r962488178

### CODE REVIEW

File: src/trait-bounds.md
Comment (MatthewPeterKelly) [inline src/trait-bounds.md:166]: Should `t` here be `T`?

Comment (MatthewPeterKelly) [reply src/trait-bounds.md:166]: Also,  I think that `inferred because the function` is a bit easier to read than `inferred as the function`.

File: src/trait-bounds.md
Comment (MatthewPeterKelly) [inline src/trait-bounds.md:176]: In this case, would it be correct to add a statement like:

"Note that we are require to add a lifetime bound on `T` in `requires_t_outlives_a_not_implied()` because it cannot be inferred from the argument (like it was in `requires_t_outlives_a()`)."

File: src/trait-bounds.md
Comment (MatthewPeterKelly) [inline src/trait-bounds.md:186]: This provides helpful context - good idea to add it here.

**Review by MatthewPeterKelly — COMMENTED** [review]
Thanks for adding this documentation to support https://github.com/rust-lang/rust/pull/100747. I'll plan to hold off merging that PR until this one lands (so that I can add a good reference).

**Review by MatthewPeterKelly — APPROVED** [review]
Looks good to me!

**Comment by MatthewPeterKelly** [comment]
(I don't have write access to the repo, so I think you'll need someone else to rubber stamp the PR).

**Comment by MatthewPeterKelly** [comment]
@Havvy - would you be able to do a quick review for this PR?

File: src/trait-bounds.md
Comment (Havvy) [inline src/trait-bounds.md:169]: The inputs and called parameters.

**Review by Havvy — CHANGES_REQUESTED** [review]
Reading through this, I feel like the prose is unnecessarily vague. And in fact, it fails to answer two key questions:

1. What are all the places that lifetime bounds are inferred? As written, it looks like it is only for generics in functions. Are there others?
2. How are the bounds chosen by looking at the types of the function parameters and output? In the simplest case, as shown by the example, a top level reference to a generic type gives said generic type a lifetime bound of the lifetime of the reference. Are there more complicated cases? If so, what are they?

--

The examples are good.

File: src/trait-bounds.md
Comment (lcnr) [reply src/trait-bounds.md:170]: how would you write this sentence then?

Comment (Havvy) [reply src/trait-bounds.md:170]: Just replace "inputs" with "parameters"

File: src/trait-bounds.md
Comment (Havvy) [inline src/trait-bounds.md:161]:
```suggestion
Lifetime bounds required for types to be well-formed are sometimes inferred.
```

File: src/trait-bounds.md
Comment (Havvy) [inline src/trait-bounds.md:170]: When does it add implied bounds? You have a lot of good examples, but the actual rules for how the bounds are inferred are missing. I don't actually know the rules, so I can't be helpful here. If you don't either, I guess we can merge this and open an issue or we can find somebody who does know the rules. I do imagine it would be something like "For each parameter type and output type, if the type contains a reference to a generic type `T` (declared by the function item?), then the bounds with a lifetime `'a`, then `T: 'a` is inferred."

**Review by Havvy — CHANGES_REQUESTED** [review]

File: src/trait-bounds.md
Comment (lcnr) [reply src/trait-bounds.md]: > I guess we can merge this and open an issue

let's do that then. I think explaining this would require the reference to have a section about well-formedness requirements and I don't have the time to write that one rn. Lets open an issue and i can try to look into that one at a later point.

File: src/trait-bounds.md
Comment (Havvy) [inline src/trait-bounds.md:170]: Still need to write this without referring to Rust directly. Everything is about Rust.

File: src/trait-bounds.md
Comment (Havvy) [inline src/trait-bounds.md:191]: Missing newline between sentences.

**Review by Havvy — CHANGES_REQUESTED** [review]
Couple nits.

**Comment by lcnr** [comment]
done :+1:

**Comment by lcnr** [comment]
:wave: do you think we can go ahead and merge this?

**Review by ehuss — APPROVED** [review]
Sure, I think this can go ahead and get merged. I'd agree with Havvy's comments that this seems a little informal, but I can understand how being more precise would take a ton more work.

**Pull request merged**
